ROS Node Integration

Jmeyer1292 opened the pull request:
This PR:
 * moves the stand-alone keyence library into a subdirectory (much like moveit structures its code).
 * Implements a ROS node that queries the Keyence controller at the program configured rate for more sensor data.
 * Adds a ROS service that allows for remote changing of the active program (e.g. to turn the sensor off).
 * Adds a new `GetSettings` command to the keyence library that is used to fetch info about the active program for use in the ROS node.

The ROS side of things based on previous work for the Godel project done by @gavanderhoorn and Matthew West.

@Levi-Armstrong You looked at my last PR, would you be so kind to do this one too?

Levi-Armstrong commented on keyence_library/include/keyence/impl/keyence_tcp_client.h:
```diff
@@ -17,8 +17,8 @@ class TcpClient : public Client
-  virtual int send(const void* data, std::size_t size); // override
-  virtual int recv(void* data, std::size_t size); // override
+  virtual int send(const void* data, std::size_t size) override;
+  virtual int recv(void* data, std::size_t size) override;
```
Should the virtual be removed?

Jmeyer1292 replied:
+1

Levi-Armstrong commented on CMakeLists.txt:
```diff
@@ -3,28 +3,41 @@ project(keyence_experimental)
-# Build test program for single keyence profile
-add_executable(keyence_get_profile src/utilities/get_profile.cpp)
-target_link_libraries(keyence_get_profile keyence_impl)
+add_subdirectory(keyence_library)
```
Should there be an install step?

Jmeyer1292 replied:
I don't know too much about how the install procedures work, and my only dependency, libsocket, doesn't have a debian.

I'm still happy to provide install targets.

gavanderhoorn replied:
Isn't [this](http://packages.ubuntu.com/trusty/libs/libsocket++1) the `libsocket` that you are using? It would seem there is a binary release for at least Trusty, so perhaps there is also one available for Xenial.

Edit: there seems to be a Xenial release as well ([here](http://packages.ubuntu.com/xenial/libs/libsocket++1)).

Jmeyer1292 replied:
It's not the same library, unfortunately.

Levi-Armstrong replied:
Is the plan to release this eventually?

Jmeyer1292 replied:
I don't know if we have a policy that everything should get released, but I have no incentive to do so at the moment.

Levi-Armstrong commented on keyence_library/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+
+# Build test program for single keyence profile
+add_executable(keyence_get_setting src/get_setting.cpp)
+target_link_libraries(keyence_get_setting keyence_impl)
```
same as above.

Levi-Armstrong commented on keyence_library/include/keyence/impl/messages/get_setting.h:
```diff
@@ -0,0 +1,49 @@
+  {
+    typedef Response response_type;
+    const static uint32_t size = 16;
+    const static uint8_t command_code = 0x31;
```
I believe these two should be ALL_CAPS.

Jmeyer1292 replied:
This is one case where I disagree with the style guide. I'm slightly uncomfortable defining something like `size` as `SIZE` because it could possibly collide with a C MACRO from some other header.

Plus I'd have to change all of the other commands to match. If we want to do that, I'd prefer to address it in another PR.

Levi-Armstrong commented on keyence_library/include/keyence/impl/settings_defs.h:
```diff
@@ -0,0 +1,56 @@
+  static const uint8_t freq_16000hz = 11;
+  static const uint8_t freq_32000hz = 12;
+  static const uint8_t freq_64000hz = 13;
+};
```
I believe all const variables should be ALL_CAPS.

Levi-Armstrong commented on keyence_library/src/get_setting.cpp:
```diff
@@ -0,0 +1,68 @@
+  }
+
+  return 0;
+}
```
I noticed that std::cerr and std::cout was used, should ros messaging be used?

Jmeyer1292 replied:
So the package is divided into a ROS-independent comms library (found in `keyence_library`) and a ROS front-end. This sample is part of the former, and can be built without ROS. 

Inside the library itself, you will not find any use of `cout` or `cerr`. Error reporting is conducted through exceptions and overloads are provided for use with any stream operators including `ROS_X_STREAM` commands.

Levi-Armstrong commented on src/keyence_driver_node.cpp:
```diff
@@ -0,0 +1,373 @@
+uint8_t getActiveProgramNumber(keyence::TcpClient& client)
+{
+  uint8_t level = 0; // write settings area
+  uint8_t type = 0x10;
```
Should these be moved to const variables?

Jmeyer1292 replied:
They probably should be.
+1

Jmeyer1292 commented:
@Levi-Armstrong I have:
 - Updated the readme to better reflect the new state of the code
 - Moved a lot of constants into the settings header (and they even have comments)
 - Improved some of the error handling
